### PR TITLE
[CDAP-19226] convert logviewer test

### PR DESCRIPTION
**Pull request by sumengwang**
# [CDAP-19226] convert logviewer test

## PR Type
- [ ] Bug Fix
- [ ] Feature
- [ ] Build Fix
- [x] Testing
- [x] General Improvement
- [ ] Cherry Pick

## Links
Jira: [CDAP-19226](https://cdap.atlassian.net/browse/CDAP-19226)

## Test Plan
local test and github action test passes

[CDAP-19226]: https://cdap.atlassian.net/browse/CDAP-19226?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/BeforeActions.java [hunk 1]
```diff
@@ -31,4 +31,10 @@ public void loginIfRequired() throws IOException {
     logger.info("-----------------Logging in if required------------------");
     Helper.loginIfRequired();
   }
+
+  @Before(order = 0)
+  public void createTmpDirectory() throws IOException {
```
Comment (sumengwang) [inline src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/BeforeActions.java]: This method creates a tmp dir in the runner with files in the fixtures that will be used by the pipelines

File: src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/Logviewer.java [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+  @Then("Log level popover should not show")
+  public void logLevelPopoverShouldHide() {
+    WaitHelper.waitForElementToBeHidden(Helper.locateElementByCssSelector(
+      Helper.getCssSelectorByDataTestId("log-viewer-log-level-popover")
+    ));
+  }
+
+  @Then("Scroll up to center")
+  public void scrollUpToCenter() {
+    JavascriptExecutor js = (JavascriptExecutor) SeleniumDriver.getDriver();
```
Comment (sumengwang) [inline src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/Logviewer.java:183]: I don't know how to scroll a container other than using jsexecutor

File: src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java [hunk 3]
```diff
@@ -217,4 +271,42 @@ public static void setCdapTheme(String themePath) {
       Assert.fail(e.getMessage());
     }
   }
+
+  public static void takeScreenShot() {
```
Comment (sumengwang) [inline src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java]: can take screenshots now for debugging locally. Maybe we can try to save error screenshots into the build debug files in github actions in the future

Comment (de-lan) [reply src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java]: Could we consolidate these two? https://github.com/cdapio/cdap-ui/blob/develop/src/e2e-test/java/io/cdap/cdap/ui/utils/Commands.java#L314

Comment (sumengwang) [reply src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java]: ooo, didn't see it

Comment (mrahanjam) [reply src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java]: FYI, you can add the reports/screenshots to the build debug files by giving the workflow file their corresponding directory here:
https://github.com/cdapio/cdap-ui/blob/da96db321b5e1621ab89481508265afa906889bf/.github/workflows/build.yml#L95

File: src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/Logviewer.java (hunk 2 above)
Comment (mrahanjam) [reply src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/Logviewer.java:183]: Have you looked at the helper methods in here:
https://github.com/cdapio/cdap-e2e-tests/blob/727a948be8abdad2dd8d32900c606285195b0e30/src/main/java/io/cdap/e2e/utils/ElementHelper.java#L38

Comment (sumengwang) [reply src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/Logviewer.java:183]: Yea, the issue is the behavior here is not "scrolling to an element", it kinda just wants to scroll for a fixed length, cypress had the ability to scroll to 'center', 'bottom', but I don't think selenium can do that

Comment (mrahanjam) [reply src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/Logviewer.java:183]: How do you decide what the fixed length is?

Comment (sumengwang) [reply src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/Logviewer.java:183]: experiments and estimations

Comment (mrahanjam) [reply src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/Logviewer.java:183]: I did a quick search and found this:
```javascript
js.executeScript("arguments[0].scrollIntoView({behavior: "smooth", block: "center", inline: "nearest"}))", elem);
```

have you tried it?

Comment (sumengwang) [reply src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/Logviewer.java:183]: Let me try it out

File: src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java [hunk 4]
```diff
@@ -217,4 +271,42 @@ public static void setCdapTheme(String themePath) {
...
+    } catch (IOException e) {
+      logger.info("cannot take a screenshot");
+    }
+  }
+
+  public static void waitSeconds() {
+    waitSeconds(1);
+  }
+
+  public static void waitSeconds(long second) {
```
Comment (mrahanjam) [inline src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java:275]: why do we need this? we better not use `Thread.sleep` .

Comment (sumengwang) [reply src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java:275]: There are some loading happening

Comment (mrahanjam) [reply src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java:275]: Can't you wait for an element to appear (indicating the wait is over)?

Comment (sumengwang) [reply src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java:275]: I will change

File: src/e2e-test/fixtures/airports.csv [hunk 5]
```diff
@@ -0,0 +1,10 @@
+iata,name,city,state,country,latitude,longitude
+00M,Thigpen,Bay Springs,MS,USA,31.95376472,-89.23450472
+00R,Livingston Municipal,Livingston,TX,USA,30.68586111,-95.01792778
+00V,Meadow Lake,Colorado Springs,CO,USA,38.94574889,-104.5698933
+01G,Perry-Warsaw,Perry,NY,USA,42.74134667,-78.05208056
+01J,Hilliard Airpark,Hilliard,FL,USA,30.6880125,-81.90594389
+01M,Tishomingo County,Belmont,MS,USA,34.49166667,-88.20111111
+02A,Gragg-Wade,Clanton,AL,USA,32.85048667,-86.61145333
+02C,Capitol,Brookfield,WI,USA,43.08751,-88.17786917
+02G,Columbiana County,East Liverpool,OH,USA,40.67331278,-80.64140639
```
Comment (de-lan) [inline src/e2e-test/fixtures/airports.csv]: newline missing?

File: src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/Logviewer.java [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    Assert.assertEquals(false,
+      Helper.locateElementByCssSelector(Helper.getCssSelectorByDataTestId("scroll-to-latest")).isEnabled()
+    );
+  }
+
+  @Then("Scroll to latest should be enabled")
+  public void scrollTolatestShouldBeEnabled() {
+    WaitHelper.waitForElementToBeEnabled(
+      Helper.locateElementByCssSelector(Helper.getCssSelectorByDataTestId("scroll-to-latest")));
+    Assert.assertEquals(true,
```
Comment (de-lan) [inline src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/Logviewer.java]: Could use `Assert.assertTrue` instead?

File: src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/Logviewer.java [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+      Helper.locateElementByCssSelector(Helper.getCssSelectorByDataTestId("scroll-to-latest")).isEnabled()
+    );
+  }
+
+  @Then("Advanced logs should exist")
+  public void advancedLogsShouldExist() {
+    Assert.assertEquals(true,
```
Comment (de-lan) [inline src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/Logviewer.java]: Same about `Assert. assertTrue` with the rest lines

File: src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java (hunk 4 above)
Comment (sumengwang) [reply src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java:275]: actually some of the waits are designed to wait for new data refresh. These cases do not have new elements to wait for

Comment (de-lan) [reply src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java:275]: Agree that we don't wait arbitrary time, instead wait for a target element to show up

File: src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/Logviewer.java (hunk 2 above)
Comment (sumengwang) [reply src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/Logviewer.java:183]: no this doesn't work

**Review by de-lan — APPROVED** [review]

**Review by radhikav1 — APPROVED** [review]

**Pull request merged**